RFCs/codepush-for-mobile-releases.md

```diff
@@ -0,0 +1,75 @@
+## Proposal
+
+Use [codepush](https://microsoft.github.io/code-push/) ([github](https://github.com/microsoft/react-native-code-push)) for
+- QA-ing PRs before they are merged, even by non-engineers, and
+- doing faster, leaner releases, more often, and closer to `main`.
+
+## Reasoning
+
+Codepush can be very beneficial to us in two ways.
+
+- Allow us to test/QA PRs before they are merged, and by non-engineers too.
+
+Until now, reviewers of a PR can try to run the app on their machine, but this doesn't happen often, as it is a hassle. Also, before merging, is it next to impossible for designers and PMs and anyone else non-engineer to test the PR, since they would have to set up their machine for that.
+
+As a result, we end up merging first, and then QA-ing by building a beta. The beta then can be used by non-engineers to try out the changes, and often we need to make another PR to address some issues big or small.
+
+Why not avoid all that, and allow ourselves and everyone else at artsy with the beta app, to try out changes on a PR, without the need to merge first and build a new beta?
+
+- Allow us to do faster releases, more often, and closer to `main`.
+
+They way `force` does releases is a thing to admire. `eigen` is always a bit slow with releases, and for a reason. And that reason is because it's slow to do releases. You see the issue here?
+
+From the second we decide to do a release (after we have QA-ed and everything), a couple of days at least need to pass, for the appstore review. That causes us to be extra careful, because if a bug causes a huge issue, the only way to fix this is by fixing and then waiting two more days, and email Apple to speed this process up. It causes us to be more afraid to break things, because they would remain broken for a long time.
+
+If we have a way to do hotfixes or even regular feature releases and changes in a matter of seconds or minutes, then it is much much easier to move faster. If something breaks, the fix will be delivered in a few minutes to every user. If we forgot to add something, toggle a flag, change a color, whatever, it can happen in a few minutes.
+
```

Comment:
In reality, I don't think this is true.

The only really safe time to check for an updated JS bundle is on app launch (otherwise you risk a mismatch between code and app state). This means that it will take a while for users to actually receive a hotfix, either because (a) the app is running in the background and doesn't get launched for a while, or (b) they're just not launching the app in the first place. 

Meanwhile app store review time is between a few hours and a day, after which time you have a pretty good guarantee (although admittedly not 100%) that users will be updated to the new, hotfixed version. So in reality the improvement that Codepush brings under this scenario is minimal.

Reply:
I think this is totally the other way.

appstore review times, even if they are fast, do not guarantee a user will get it. the phones usually update be themselves when they are charging. could during the night or in the regular day. some people have the auto update off. there are definitely reasons why the update would not be so immediate.

with codepush, we can check whenever, my proposal would be to check every appstart, probably also every app foregrounding. if a hotfix is out, then it will be downloaded at appstart or fg, and applied on the next start. very easy. also, like I said, we can control that, so if there is a major bug, we can choose to apply it immediately when it's downloaded. yes, it might be jarring to have one flash one time, but better than a huge huge bug. (I would not force the update directly if not huuuge bug.)

there is no issue with app state at all in either case. the app will do migrations as normal, if we have a migration for the update. there is no difference on the apostate if it's a native or a OTA update.

Reply:
> First, [..] it just takes a while for apple/google to review releases and there isn't much we can do about it.

But we can. we can use codepush if we want, to speed things up.

> Second, Apple and Google both don't like shipping regular updates without them knowing about it [..]

Which is why I think its good to do regular appstores updates every few releases. Just to keep things ok for everyone.

> Third, [..] each app launch will result in the app restarting after the new bundle is downloaded, for everyone. [..] especially knowing that the jsbundle is about 8mb now (7.9mb).

I think we are not all on the same page on how this works. Firstly, the size is ok. The regular app on the appstore is 45mb, so an update of 45 and an update of 8 is definitely bigger, if that matters. Secondly, it will not restart the app. As I proposed, the app would download the update, but only apply it on the next appstart, which will not affect the user experience at all.

I think we should start using it for the PR checks first, and once people are familiar on how this works, then we can move to possibly some hotfixes. And from there we can decide if we want to use it for more, or keep it there.

Reply:
Agree, let's give it a try. I am excited about it

Reply:
> I think we are not all on the same page on how this works

I understand that we can decide when we want the app bundle to be updated and when we want to make the changes to take effect and this could be at the app start, on AppState change or at the next session, this is clear to me. In most cases I saw though, companies usually tend to do it at the app start since they're trying to make sure users get a fix asap, and if a fix is available, it would wait to get the latest js bundle on app launch and start the app and reloads the splash screen, and if mid-session, it restarts the app if specified (according to code-push docs https://docs.microsoft.com/en-gb/appcenter/distribution/codepush/rn-api-ref#installmode).

If we use code-push for normal app updates, then yes, 8mb is definitely a big advantage compared to 45mb.

> I think we should start using it for the PR checks first, and once people are familiar on how this works, then we can move to possibly some hotfixes

I agree with this, let's try it out. A quick suggestion in this case where we have an RFC for a technology that we might not be all familiar with, is to also add a small paragraph or description for what the tool is and what it does, and if needed how it works. This could help you gather more feedback about it